### ISSUE TITLE
fluid下有类似tf.where()接口么？

### ISSUE THREAD

**Issue by luameows** [PaddlePaddle/Paddle#13200]
我想实现amsoftmax这样一个损失函数，其中需要调用类似`tf.where(condition,x,y)`的接口。该函数作用是根据condition返回x或y中的元素。
请问fluid下是否有相关接口函数呢？
如果没有，能否通过numpy操作实现？

**Comment by luotao1** [comment]
fluid没有完整的where接口。相关接口可以参考：
- gatter: http://www.paddlepaddle.org/documentation/api/zh/0.14.0/layers.html#gather
- scatter: http://www.paddlepaddle.org/documentation/api/zh/0.14.0/layers.html#scatter
- less_than: http://www.paddlepaddle.org/documentation/api/zh/0.14.0/layers.html#less-than

**Comment by luameows** [comment]
@luotao1 您提供的这些接口对我而言用不上，并且scatter是最近才封装的，[#12725](https://github.com/PaddlePaddle/Paddle/issues/12725) 。我有两个问题想请教下。
1.对于[#12725](https://github.com/PaddlePaddle/Paddle/issues/12725) 更新的接口，我如果要采用该版本，应该如何更新我的paddle呢？之前尝试过pip install，发现并没有更新。
2.fluid每一层运算后的返回值`Variable`或`tensor`，是numpy形式么？能否转成numpy形式？我能否以numpy数组运算的方式来自己编写类似`tf.where`这样的功能？

**Comment by luotao1** [comment]
> 如何更新我的paddle

http://www.paddlepaddle.org/documentation/docs/zh/0.13.0/build_and_install/pip_install_cn.html#id6 可以选择最新的whl包进行安装。

> fluid每一层运算后的返回值Variable或tensor，是numpy形式么

是的。可以参考 https://github.com/PaddlePaddle/Paddle/issues/13029#issuecomment-416574599 打印。

> 自己编写类似tf.where这样的功能

我们有一个[test_if_else_op.py](https://github.com/PaddlePaddle/Paddle/blob/9557cc218d3d71947d7204dce6d711126eb80ad0/python/paddle/fluid/tests/test_if_else_op.py)，可供参考。

**Comment by luameows** [comment]
@luotao1 看了下您提供的这个例程，只是比较了`IFELSE`与numpy操作的结果，并且在numpy方法中，所有变量均是通过numpy定义得到，而没有关于numpy操作layer的方式呢。
我现在的目的是：
```
ipt=layers.data(name='data', shape=[1], dtype='float32')
# 如何将ipt转化为可操作的numpy.nparray形式。

# numpy具体操作.....得到output

#如何将output转化为可以输入fc或loss的形式
loss = layers.cross_entropy(input=output, label=label)
```


另外，关于您提供的`ifelse`接口，API上没有具体说明适用方法。对于下面这个例子，请问`input`与`output`得到的分别是什么呢？最后返回`ie()`是将两个block的结果合并？
![image](https://user-images.githubusercontent.com/30432086/45020255-0959da80-b061-11e8-969f-a794614c8efc.png)

**Comment by luotao1** [comment]
> 而没有关于numpy操作layer的方式呢

不能使用numpy来操作layer。这样反向的信息无法保存。

> 关于您提供的ifelse接口，API上没有具体说明适用方法

API上没有这个接口，只是一个测试用例。

> 最后返回ie()是将两个block的结果合并

是的。

**Comment by luameows** [comment]
@luotao1 谢谢。
我通过另外的计算方法绕过了`tf.where()`的问题。但是又出现了几个新的情况，还请帮忙看一下。
1.提取特征feature，shape为[-1，512]，之后通过`layers.l2_normalize(feature,axis=1,epsilon=1e-10)`进行归一化之后，所有值都变成nan。已经确认feature值为正常。

2.创建学习参数w，发现所有值都是nan
```
fw_param_attrs=fluid.ParamAttr(learning_rate=1,
                                    initializer=fluid.initializer.Xavier())
    fw=fluid.layers.create_parameter(shape=[embedding_dim,class_dim],
                                    dtype='float32',
                                    name='feature_weights',
                                    attr=fw_param_attrs)
```

3.通过numpy创建学习参数w并归一化，之后利用`layers.assign`将其转化为`variabel`，该方法是不是无法保存反向传播信息？

**Comment by luameows** [comment]
1.更新下，我又尝试通过其他接口，自己构造l2_norm，代码如下。
```
input_x_square=fluid.layers.square(x=out1)
input_x_sum=fluid.layers.reduce_sum(input=input_x_square,dim=1,keep_dim=True)
input_x_sqrt=fluid.layers.sqrt(x=input_x_sum)
input_x_norm=fluid.layers.elementwise_div(x=out1,y=input_x_sqrt,axis=0)
```
跟踪上诉变量，发现`input_x_sqrt`及之前部分均正常。当执行`elementwise_div`，所有值全部变成nan。


2.关于`create_parameter`初始化全为nan，还是没有找到原因及解决办法。

**Comment by luotao1** [comment]
`axis=0`是否正确？`axis`默认=-1

**Comment by luotao1** [comment]
关于l2_normalize的问题，请问您用的是什么版本的whl包。0.14.0的whl包（包含 #11348 ）已经修复过一次。如果您还出现问题，请提供复现该问题的case。比如输入feature map是什么，等。

**Comment by luameows** [comment]
@luotao1 
1.关于axis的值，官方介绍表示是从第几个维度开始。我也尝试过其他的值如0，-1，结果都是一样出现了问题
![image](https://user-images.githubusercontent.com/30432086/45070962-8f7c2c80-b106-11e8-9098-fd9cef504300.png)
2.关于whl包。我也怀疑是不是这个问题。现在paddle是科赛网上工作人员安装的，通过pip install paddlepaddle-gpu ==0.14.0，请问下我如何查看当前版本呢？

**Comment by lucywsq** [comment]
您好，此issue在近一个月内暂无更新，我们将于今天内关闭。若在关闭后您仍需跟进提问，可重新开启此问题， 我们将在24小时内回复您。因关闭带来的不便我们深表歉意，请您谅解~感谢您对PaddlePaddle的支持!
